File: src/vision_node/vision_node/face_detector_node.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int32, Float64, String
from ament_index_python.packages import get_package_share_directory  
import time
import os
import cv2 as cv
import threading

class FaceDetectorNode(Node):
    def __init__(self):
        super().__init__('face_detector')

        self.publisher_1 = self.create_publisher(Int32, '/vision/face_count', 10)
        self.publisher_2 = self.create_publisher(Float64, '/vision/fps', 10)
        self.publisher_3 = self.create_publisher(String, '/vision/alerts', 10)

        #open webcam using opencv
        self.cap = cv.VideoCapture(0)
        self.latest_frame = None
        self.frame_lock = threading.Lock()

        #capture thread runs independently
        self.capture_thread = threading.Thread(
            target=self.capture_loop, daemon=True)
        self.capture_thread.start()



        #load haar cascade from 'haarcascade_frontalface_default.xml'
        cascade_path = os.path.join(
            get_package_share_directory('vision_node'),
            'data',
            'haarcascade_frontalface_default.xml'
        )
        self.face_cascade = cv.CascadeClassifier(cascade_path)

        if self.face_cascade.empty():
            self.get_logger().error(
                f'Failed to load cascade from: {cascade_path}')
            raise RuntimeError('Cascade classifier failed to load. Check data_files in setup.py.')


        self.prev_time = time.time()

        # create timer at 30Hz (period = 1/30)
        self.timer =  self.create_timer(1/30, self.process_frame)

        self.get_logger().info('FaceDetector node started')

    def capture_loop(self):
        while rclpy.ok():
            ret, frame = self.cap.read()
            if ret:
                with self.frame_lock:
                    self.latest_frame = frame

    def process_frame(self):
        #capture a frame from the webcam
        with self.frame_lock:
            if self.latest_frame is None:
                return
            frame = self.latest_frame.copy()
        
        current_time = time.time()
        fps = 1.0/(current_time - self.prev_time)
        self.prev_time = current_time

        #convert the frame to grayscale for face detection
        gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
        #detect faces in the grayscale frame using detectMultiscale
        faces = self.face_cascade.detectMultiScale(gray, scaleFactor=1.1, minNeighbors=5, minSize=(30,30))

        #draw rectangles around detected faces on the original frame
        for (x,y,w,h) in faces:
            cv.rectangle(frame, (x,y), (x+w, y+h), (0, 0, 255), 2)
            #put fps text on the frame using cv.putText
        cv.putText(frame, 'FPS: %.2f' % fps, (10,30), cv.FONT_HERSHEY_SIMPLEX, 1, (0,255,0), 2)
        cv.putText(frame, 'Faces: %d' % len(faces), (10,70), cv.FONT_HERSHEY_SIMPLEX, 1, (0,255,0), 2)

        #face count (len of face array) as Int32
        face_count_msg = Int32()
        face_count_msg.data = len(faces)
        self.publisher_1.publish(face_count_msg)
            
        #publish fps as Float64
        fps_msg = Float64()
        fps_msg.data = fps
        self.publisher_2.publish(fps_msg)

        face_detection_alert_msg = String()
        if len(faces) > 0:
            face_detection_alert_msg.data = 'FACE DETECTED'
        else:
            face_detection_alert_msg.data = 'NO FACES DETECTED'
        self.publisher_3.publish(face_detection_alert_msg)

        cv.imshow('face detection', frame)

        cv.waitKey(1)

    def destroy_node(self):
        self.cap.release()
        cv.destroyAllWindows()
        super().destroy_node()

def main(args=None):
    rclpy.init(args=args)
    node = FaceDetectorNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/vision_node/src/face_detector_node.cpp
#include <atomic>
#include <chrono>
#include <memory>
#include <mutex>
#include <string>
#include <thread>

#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/objdetect.hpp>
#include <opencv2/videoio.hpp>

#include "ament_index_cpp/get_package_share_directory.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64.hpp"
#include "std_msgs/msg/int32.hpp"
#include "std_msgs/msg/string.hpp"

class FaceDetectorNode : public rclcpp::Node
{
public:
  FaceDetectorNode()
  : Node("face_detector"), running_(true), has_frame_(false)
  {
    face_count_publisher_ = this->create_publisher<std_msgs::msg::Int32>("/vision/face_count", 10);
    fps_publisher_ = this->create_publisher<std_msgs::msg::Float64>("/vision/fps", 10);
    alert_publisher_ = this->create_publisher<std_msgs::msg::String>("/vision/alerts", 10);

    cap_.open(0);
    if (!cap_.isOpened()) {
      throw std::runtime_error("Failed to open webcam device 0");
    }

    const std::string cascade_path =
      ament_index_cpp::get_package_share_directory("vision_node") +
      "/data/haarcascade_frontalface_default.xml";

    if (!face_cascade_.load(cascade_path)) {
      throw std::runtime_error(
        "Cascade classifier failed to load. Check data files in CMakeLists.txt.");
    }

    prev_time_ = std::chrono::steady_clock::now();

    capture_thread_ = std::thread(&FaceDetectorNode::capture_loop, this);

    timer_ = this->create_wall_timer(
      std::chrono::milliseconds(33),
      std::bind(&FaceDetectorNode::process_frame, this));

    RCLCPP_INFO(this->get_logger(), "FaceDetector node started");
  }

  ~FaceDetectorNode() override
  {
    running_ = false;
    if (capture_thread_.joinable()) {
      capture_thread_.join();
    }
    cap_.release();
    cv::destroyAllWindows();
  }

private:
  void capture_loop()
  {
    while (running_ && rclcpp::ok()) {
      cv::Mat frame;
      if (cap_.read(frame) && !frame.empty()) {
        std::lock_guard<std::mutex> lock(frame_mutex_);
        latest_frame_ = frame.clone();
        has_frame_ = true;
      }
    }
  }

  void process_frame()
  {
    cv::Mat frame;
    {
      std::lock_guard<std::mutex> lock(frame_mutex_);
      if (!has_frame_) {
        return;
      }
      frame = latest_frame_.clone();
    }

    const auto current_time = std::chrono::steady_clock::now();
    const double elapsed = std::chrono::duration<double>(current_time - prev_time_).count();
    prev_time_ = current_time;
    const double fps = elapsed > 0.0 ? 1.0 / elapsed : 0.0;

    cv::Mat gray;
    cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);

    std::vector<cv::Rect> faces;
    face_cascade_.detectMultiScale(gray, faces, 1.1, 5, 0, cv::Size(30, 30));

    for (const auto & face : faces) {
      cv::rectangle(frame, face, cv::Scalar(0, 0, 255), 2);
    }

    cv::putText(
      frame, "FPS: " + std::to_string(fps), cv::Point(10, 30),
      cv::FONT_HERSHEY_SIMPLEX, 1.0, cv::Scalar(0, 255, 0), 2);
    cv::putText(
      frame, "Faces: " + std::to_string(faces.size()), cv::Point(10, 70),
      cv::FONT_HERSHEY_SIMPLEX, 1.0, cv::Scalar(0, 255, 0), 2);

    std_msgs::msg::Int32 face_count_msg;
    face_count_msg.data = static_cast<int32_t>(faces.size());
    face_count_publisher_->publish(face_count_msg);

    std_msgs::msg::Float64 fps_msg;
    fps_msg.data = fps;
    fps_publisher_->publish(fps_msg);

    std_msgs::msg::String alert_msg;
    alert_msg.data = faces.empty() ? "NO FACES DETECTED" : "FACE DETECTED";
    alert_publisher_->publish(alert_msg);

    cv::imshow("face detection", frame);
    cv::waitKey(1);
  }

  rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr face_count_publisher_;
  rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr fps_publisher_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr alert_publisher_;
  rclcpp::TimerBase::SharedPtr timer_;

  cv::VideoCapture cap_;
  cv::CascadeClassifier face_cascade_;
  cv::Mat latest_frame_;
  std::mutex frame_mutex_;
  std::thread capture_thread_;
  std::atomic<bool> running_;
  bool has_frame_;
  std::chrono::steady_clock::time_point prev_time_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FaceDetectorNode>());
  rclcpp::shutdown();
  return 0;
}
